feat: [CDS-82830]: Adding documentation for CD custom dashboards (#4629)
harness/developer-hub: docs/continuous-delivery/monitor-deployments/using-cd-custom-dashboards.md

---
title: Using Custom Dashboards to monitor CD deployments
description: Using Custom Dashboards to monitor CD deployments.
sidebar_position: 1
helpdocs_topic_id: phiv0zaoex
helpdocs_category_id: 0gyfk4938x
helpdocs_is_private: false
helpdocs_is_published: true
---

This topic describes the various Explores available for monitoring CD deployments, and also the data available in each explore. The below image shows the list of explores available for visualizing CD data.

<docimage path={require('./static/using-cd-custom-dashboards-1.png')} width="60%" height="60%" title="Click to view full size image" />

It is recommended to refer [this](../../platform/dashboards/create-dashboards.md) documentation first to get a basic overview on how to create dashboards.

## Deployments and Services

This explore can be used to visualize CD data related to pipeline executions. Each explore has a set of views. The views present in this explore are:

<docimage path={require('./static/using-cd-custom-dashboards-20.png')} width="60%" height="60%" title="Click to view full size image" />

> Let us go through each of the view listed in the above image, and describe the fields and it's usages.

Reviewer: ```suggestion
Let's go through each of the views listed above and describe its fields and usages.
```

> ### Approval Stage
> This view contains the data related to the Helm manifest info which is used in CD Stage executions. The data is recorded per CD Stage Execution, meaning if there are 10 executions of a single service which has a helm manifest linked to it this view will also have 10 records.
> 

Reviewer: ```suggestion
### Approval Stage

This view contains the data related to the Helm manifest info that is used in CD stage executions. The data is recorded per CD stage execution. For example, if there are 10 executions of a single service with a Helm manifest linked to it, this view will have 10 records.

```

<docimage path={require('./static/using-cd-custom-dashboards-21.png')} width="60%" height="60%" title="Click to view full size image" />

Filter-only Field:
1. Current Aggregation Period: 

List of dimensions:
1. End Time: End time of the stage execution
2. Stage Name: Name of the stage
3. Stage Status: Current status of the stage(`RUNNING`/`FAILED` etc.)
4. Stage Type: Type of the stage(`Approvals`)
5. Start Time: Start time of the stage execution

List of measures:
1. Stage Execution Duration: Average stage execution duration

### Deployments

> This view contains the data related to the pipeline executions. The data is recorded per pipeline execution, meaning if there are 10 pipeline executions, then this view will also have 10 records.

Reviewer: ```suggestion
This view contains the data related to pipeline executions. The data is recorded per pipeline execution, meaning if there are 10 pipeline executions, then this view will also have 10 records.
```

5. Pipeline Name: Name of the pipeline
6. Project ID: Project identifier where the pipeline got executed
7. Start Time: Start time of the pipeline execution
8. Status: Pipeline status(`RUNNING`/`ABORTED`/`SUCCESS` etc.)
9. Trigger Type: Type of trigger used to execute the pipeline(`MANUAL`/`WEBHOOK` etc.)

List of measures:
1. Change Failure Rate: Total deployments that have been reverted divided by the total no. of deployments
2. Failed Deployments: No. of deployments that failed with `FAILED`/`ABORTED` statuses
3. Failure Rate: Failure rate of deployments(Failed Deployments divided by total no. of deployments)
4. Last Deployment: Last deployment time
5. Longest Deployment: Longest deployment time
6. Mean duration: Average pipeline duration
7. Mean duration trend: This compares the mean duration trend from the previous period to the current period
8. Median Duration: Median duration of the deployment
9. Success Rate: Success rate of deployments(deployments completed with `SUCCESS` status)
10. Total Deployments: Total no. of deployments
11. Total Deployments Trend: This compares the deployment count from the previous period to the current period

### Execution Tags

> This view contains the data related to the pipeline execution tags. It will contain the tags that were present in the pipeline at the start of the execution.

Reviewer: ```suggestion
This view contains the data related to pipeline execution tags. It will contain the tags that were present in the pipeline at the start of the execution.
```

<docimage path={require('./static/using-cd-custom-dashboards-7.png')} width="60%" height="60%" title="Click to view full size image" />

List of dimensions:
1. Execution tag: The execution tag, it will be displayed as `key:value`. If there are let's say 5 tags in a pipeline, there will be 5 rows having 1 tag(`key:value` pair) in each row

### Organization Tags
This view contains the data related to the organization tags. It will contain the real-time tags of the organization.

<docimage path={require('./static/using-cd-custom-dashboards-10.png')} width="60%" height="60%" title="Click to view full size image" />

List of dimensions:

> 1. Organization tag: The organization tag, it will be displayed as `key:value`. If there are let's say 5 tags in an organization, there will be 5 rows having 1 tag(`key:value` pair) in each row

Reviewer: ```suggestion
1. Organization tag: The organization tag. It will be displayed as `key:value`. For example, if there are 5 tags in an organization, there will be 5 rows having 1 tag (`key:value` pair) in each row.
```

### Organizations
This view contains the data related to organizations. 

:::note

> This will only show the organizations, which had atleast 1 pipeline execution(in any project). It won't display the organizations which didn't contain any pipeline execution.

Reviewer: ```suggestion
This will only show the organizations that had at least 1 pipeline execution (in any project). It won't display the organizations that didn't contain any pipeline executions.
```

:::

<docimage path={require('./static/using-cd-custom-dashboards-11.png')} width="60%" height="60%" title="Click to view full size image" />

List of dimensions:

> 1. Organization tag: The organization tag, it will be displayed as `key:value`. If there are let's say 5 tags in an organization, there will be 5 rows having 1 tag(`key:value` pair) in each row

Reviewer: ```suggestion
1. Organization tag: The organization tag, displayed as `key:value`. For example, if there 5 tags in an organization, there will be 5 rows having 1 tag (`key:value` pair) in each row.
```

### Pipeline Tags

> This view contains the data related to the pipeline tags. It will contain the real-time tags of the pipeline.

Reviewer: ```suggestion
This view contains the data related to pipeline tags. It will contain the real-time tags of the pipeline.
```

:::note

> This will only show the pipelines, which had atleast 1 pipeline execution. It won't display the tags for the pipelines which didn't get executed even once.

Reviewer: ```suggestion
This will only show the pipelines that had at least 1 pipeline execution. It won't display the tags for the pipelines that didn't get executed even once.
```

:::

<docimage path={require('./static/using-cd-custom-dashboards-12.png')} width="60%" height="60%" title="Click to view full size image" />

List of dimensions:

> 1. Pipeline tag: The pipeline tag, it will be displayed as `key:value`. If there are let's say 5 tags in a pipeline, there will be 5 rows having 1 tag(`key:value` pair) in each row

Reviewer: ```suggestion
1. Pipeline tag: The pipeline tag, displayed as `key:value`. For example, if there are 5 tags in a pipeline, there will be 5 rows having 1 tag (`key:value` pair) in each row.
```

### Pipelines
This view contains the data related to the pipelines. 

:::note

> This will only show the pipelines, which had atleast 1 execution. It won't display the pipelines which didn't had any execution.

Reviewer: ```suggestion
This will only show pipelines that had at least 1 execution. It won't display the pipelines that didn't had any executions.
```

:::

<docimage path={require('./static/using-cd-custom-dashboards-13.png')} width="60%" height="60%" title="Click to view full size image" />

List of dimensions:
1. Created Time: Creation time of the pipeline
2. Modified Name: Modification time of the pipeline
3. Pipeline Name: Name of the pipeline

### Projects
This view contains the data related to the projects. 

:::note

> This will only show the projects, which had atleast 1 pipeline execution. It won't display the projects which didn't had any pipeline execution.

Reviewer: ```suggestion
This will only show projects with at least 1 pipeline execution. It won't display the projects that didn't have any pipeline executions.
```

:::

<docimage path={require('./static/using-cd-custom-dashboards-14.png')} width="60%" height="60%" title="Click to view full size image" />

List of dimensions:
1. Created Time: Creation time of the project
2. Modified Name: Modification time of the project
3. Project Name: Name of the project

### Reverted Deployments

> This view contains the data related to the pipeline executions which have been reverted.

Reviewer: ```suggestion
This view contains the data related to reverted pipeline executions.
```

<docimage path={require('./static/using-cd-custom-dashboards-15.png')} width="60%" height="60%" title="Click to view full size image" />

Filter-only Field:
1. Current Aggregation Period: 

List of dimensions:
1. Deployment Duration: Duration of the pipeline execution
2. End Time: End time of the pipeline execution
3. Is Revert Execution (Yes / No): Specifies if the execution got reverted
4. Organization ID: Organization identifier where the pipeline got executed
5. Original Execution ID: Original execution id of the pipeline which got reverted
6. Pipeline Name: Name of the pipeline
7. Project ID: Project identifier where the pipeline got executed
8. Start Time: Start time of the pipeline execution

> 9. Status: Pipeline status(`RUNNING`/`ABORTED`/`SUCCESS` etc.)
> 10. Trigger Type: Type of trigger used to execute the pipeline(`MANUAL`/`WEBHOOK` etc.)

Reviewer: ```suggestion
9. Status: Pipeline status (`RUNNING`/`ABORTED`/`SUCCESS` etc.)
10. Trigger Type: Type of trigger used to execute the pipeline (`MANUAL`/`WEBHOOK` etc.)
```

List of measures:
1. Failed Deployments: No. of deployments that failed with `FAILED`/`ABORTED` statuses
2. Failure Rate: Failure rate of deployments(Failed Deployments divided by total no. of deployments)
3. Last Deployment: Last deployment time
4. Longest Deployment: Longest deployment time
5. Mean duration: Average pipeline duration
6. Mean duration trend: This compares the mean duration trend from the previous period to the current period
7. Mean Time to Restore: Measures the mean time to restore
8. Median Duration: Median duration of the deployment

> 9. Success Rate: Success rate of deployments(deployments completed with `SUCCESS` status)

Reviewer: ```suggestion
9. Success Rate: Success rate of deployments (deployments completed with `SUCCESS` status)
```

10. Total Deployments: Total no. of deployments
11. Total Deployments Trend: This compares the deployment count from the previous period to the current period

### Runtime Inputs
This view contains the data related to the runtime inputs provided at the time of the pipeline execution. This will contain only the user provided runtime inputs when a pipeline is executed.

:::note

> 1. This doesn't show nested runtime inputs. For e.g. let's say a service has a variable as runtime input, and a pipeline has service also as runtime. Then we won't show the service variable in the list
> 2. This doesn't store execution inputs. 

Reviewer: ```suggestion
1. This doesn't show nested runtime inputs. For example, let's say a service has a variable as runtime input, and a pipeline has its service as a runtime also. Then Harness won't show the service variable in the list.
2. This doesn't store execution inputs. 
```

List of dimensions:
1. Artifact Display Name: Display name of the artifact being deployed in the execution
2. Artifact Tag: Artifact tag being deployed
3. Deployment Type: Deployment type of the service
4. Environment Name: Name of the environment being used for the execution
5. Environment Type: Type of the environment(`Production`/`PreProduction`)
6. Rollback Duration: Rollback duration of the stage
7. RolledBack (Yes / No): Signifies if the stage execution did roll back
8. Service Name: Name of the service being deployed.
9. Service Stage End Time: End time of the CD stage
10. Service Stage Start Time: Start time of the CD stage
11. Service Stage Status: Status of the CD stage

List of measures:
1. Failed Service Deployments: No. of deployments that failed with `FAILED`/`ABORTED` statuses
2. Failure Rate: Failure rate of deployments(Failed Deployments divided by total no. of deployments)
3. Service Count: Distinct no. of services that got deployed
4. Success Rate: Success rate of the deployment
5. Successful Service Deployments: No. of executions that completed with `SUCCESS` status.
6. Total Service Deployments: Total no. of deployments

### Users

> This view contains the data related to the users. 

Reviewer: ```suggestion
This view contains the data related to users. 
```

:::note

> This will only show the users, which have triggered atleast 1 pipeline execution. It won't display the users which didn't trigger any pipeline execution.

Reviewer: ```suggestion
This will only show the users that have triggered at least 1 pipeline execution. It won't display the users that didn't trigger any pipeline execution.
```

:::

<docimage path={require('./static/using-cd-custom-dashboards-19.png')} width="60%" height="60%" title="Click to view full size image" />

List of dimensions:
1. Created At: Time at which user was created
2. Email id: Email id of the user
3. Updated At: Time at which user was updated
4. User Name: User name

## Deployments and Services V2 (Behind FF)

> This explore can be used to visualize CD data related to pipeline, stage and step executions. Each explore has a set of views. The views present in this explore are:

Reviewer: ```suggestion
This explore can be used to visualize CD data related to pipeline, stage, and step executions. Each explore has a set of views. The views present in this explore are:
```

<docimage path={require('./static/using-cd-custom-dashboards-2.png')} width="60%" height="60%" title="Click to view full size image" />

Let us go through each of the view listed in the above image, and describe the fields and it's usages.

### CD Stage Execution Helm Manifest Info

> This view contains the data related to the Helm manifest info which is used in CD Stage executions. The data is recorded per CD Stage Execution, meaning if there are 10 executions of a single service which has a helm manifest linked to it this view will also have 10 records.

Reviewer: ```suggestion
This view contains the data related to the Helm manifest info that is used in CD Stage executions. The data is recorded per CD stage execution, meaning if there are 10 executions of a single service that has a Helm manifest linked to it this view will also have 10 records.
```

<docimage path={require('./static/using-cd-custom-dashboards-3.png')} width="60%" height="60%" title="Click to view full size image" />

List of dimensions:
1. Chart Version: Chart version of the manifest
2. Helm Version: Helm version of the manifest
3. Manifest ID: Manifest id configured in the service
4. Manifest Type: Type of the manifest, currently it only stores for manifest of type `HelmChart`

### CD Stage Executions

> This view contains the data related to the CD Stage executions. The data is recorded per CD Stage Execution, meaning if there are 10 pipeline executions, which contain a single CD stage, then this view will also have 10 records.

Reviewer: ```suggestion
This view contains data related to the CD stage executions. The data is recorded per CD stage execution, meaning if there are 10 pipeline executions that contain a single CD stage, then this view will also have 10 records.
```
